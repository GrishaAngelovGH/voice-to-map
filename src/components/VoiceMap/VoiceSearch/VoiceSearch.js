import { useState, useEffect, Fragment } from 'react'
import { useDispatch } from 'react-redux'
import { toast } from 'react-toastify'

import Fab from '@mui/material/Fab'
import TextField from '@mui/material/TextField'

import recognition from "../recognition"
import { addLocation } from '../../../slices/locationSlice'

const VoiceSearch = () => {
    const [isRecoding, setIsRecording] = useState(false)
    const [detectedInput, setDetectedInput] = useState('')
    const dispatch = useDispatch()

    const handleClick = () => {
        if (!isRecoding) {
            recognition.start()
            setIsRecording(true)
        }
    }

    useEffect(() => {
        recognition.onresult = event => {
            const input = event.results[0][0].transcript

            setDetectedInput(input)
            dispatch(addLocation(input))
        }

        recognition.onerror = ({ error }) => {
            if (error === 'no-speech') {
                setDetectedInput('No detected speech')
                toast.error('No detected speech', { theme: 'colored' })
            }
            recognition.stop()
            setIsRecording(false)
        }

        recognition.onspeechend = () => {
            recognition.stop()
            setIsRecording(false)
        }
    }, [dispatch])

    return (
        <Fragment>
            <div className="col-md-3">
                <Fab className="mt-1 mb-1" variant="extended" color={!isRecoding ? "inherit" : "warning"} onClick={handleClick}>
                    {!isRecoding ? "Record Voice" : "Recording..."}
                </Fab>
            </div>
            <div className="col-md-4 bg-light p-3 rounded">
                <TextField label="Detected Input" variant="outlined" color='info' value={detectedInput} disabled fullWidth={true} />
            </div>
        </Fragment>
    )
}

export default VoiceSearch