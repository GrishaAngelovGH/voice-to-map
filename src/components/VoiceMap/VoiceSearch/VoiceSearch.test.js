import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'

import VoiceSearch from './VoiceSearch'

const mockStore = configureStore()
const store = mockStore({ locations: { lastLocation: '', history: [] } })

test('should render VoiceSearch component', () => {
    const view = render(
        <Provider store={store}>
            <VoiceSearch />
        </Provider>
    )

    expect(view).toMatchSnapshot()
})