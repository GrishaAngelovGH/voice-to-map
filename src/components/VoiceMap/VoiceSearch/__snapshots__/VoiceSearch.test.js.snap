// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render VoiceSearch component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="col-md-4"
      >
        <button
          class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeLarge MuiFab-colorInherit mt-1 css-16pfgg4-MuiButtonBase-root-MuiFab-root"
          tabindex="0"
          type="button"
        >
          Record Voice
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
      <div
        class="col-md-4 bg-light p-3 rounded"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorInfo Mui-disabled css-rp6sv4-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Detected Input
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorInfo Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl css-1udmear-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              disabled=""
              id=":r0:"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Detected Input
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="col-md-4"
    >
      <button
        class="MuiButtonBase-root MuiFab-root MuiFab-extended MuiFab-sizeLarge MuiFab-colorInherit mt-1 css-16pfgg4-MuiButtonBase-root-MuiFab-root"
        tabindex="0"
        type="button"
      >
        Record Voice
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="col-md-4 bg-light p-3 rounded"
    >
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorInfo Mui-disabled css-rp6sv4-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r0:"
          id=":r0:-label"
        >
          Detected Input
        </label>
        <div
          class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorInfo Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl css-1udmear-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiOutlinedInput-input MuiInputBase-input Mui-disabled css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled=""
            id=":r0:"
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-1ftyaf0"
            >
              <span>
                Detected Input
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
